packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts

```diff
@@ -0,0 +1,52 @@
+import { NextFunction } from "express"
+import { MedusaRequest, MedusaResponse } from "../../../../types/routing"
+import { StoreCreateCartType } from "../../../store/carts/validators"
+
+/**
+ * If a publishable key (PK) is passed in the header of the request AND
+ * the request carries a sales channel id param in the url or body,
+ * we check if the sales channel is valid for the key.
+ *
+ * If the request does not carry a sales channel id, we attempt to assign
+ * a sales channel associated with the PK.
+ *
+ * @throws If sales channel id is passed as a url or body param
+ *         but that id is not in the scope defined by the PK from the header.
+ *         If the PK is associated with multiple sales channels but no
+ *         sales channel id is passed in the request.
+ */
+export async function ensurePublishableKeyAndSalesChannelMatch(
+  req: MedusaRequest<StoreCreateCartType> & { publishableApiKeyScopes },
+  res: MedusaResponse,
+  next: NextFunction
+) {
+  const pubKey = req.get("x-publishable-api-key")
+
+  if (pubKey) {
+    const pubKeySalesChannels =
+      req.publishableApiKeyScopes?.sales_channel_ids ?? []
+    const channelId = req.validatedBody?.sales_channel_id
+
+    req.errors = req.errors ?? []
+
+    if (pubKeySalesChannels.length) {
+      if (channelId && !pubKeySalesChannels.includes(channelId)) {
+        req.errors.push(
+          `Sales channel ID in payload ${channelId} is not associated with the Publishable API Key in the header.`
+        )
+      }
+
+      if (!channelId) {
+        if (pubKeySalesChannels.length > 1) {
```

Comment:
This should not be allowed in the `api-key` module, right? I think it's fine to have the check but in theory it should never happen?

Reply:
Not 100% sure what you mean. The sales channels for the publishable key are retrieved via the link and attached to the scopes in the other middleware in the PR. These are not from the API key module.

Reply:
@olivermrbl Ah sorry I wasn't clear, what I meant is when you associate an api key with a sales channel, we shouldn't allow you to have 2 sales channels assigned to the same publishable key. I incorrectly referenced the API key module, my bad.